Require tls termination in route tls configuration

kargakis opened the pull request:
Forked from https://github.com/openshift/origin/pull/6056#discussion_r45864181

@openshift/api-review

kargakis commented:
Also @pweil- @ramr

deads2k commented:
@jwforres

deads2k commented on pkg/route/api/v1beta3/types.go:
```diff
@@ -75,7 +75,7 @@ type RouterShard struct {
 type TLSConfig struct {
-	// Termination indicates termination type.  If termination type is not set, any termination config will be ignored
+	// Termination indicates termination type.
 	Termination TLSTerminationType `json:"termination,omitempty"`
```
This should not be `omitempty`

kargakis replied:
oops, missed. updated now

deads2k commented on pkg/route/api/validation/validation.go:
```diff
@@ -109,7 +109,7 @@ func validateTLS(route *routeapi.Route) fielderrors.ValidationErrorList {
 		}
 	default:
-		msg := fmt.Sprintf("invalid value for termination, acceptable values are %s, %s, %s, or empty (no tls specified)", routeapi.TLSTerminationEdge, routeapi.TLSTerminationPassthrough, routeapi.TLSTerminationReencrypt)
+		msg := fmt.Sprintf("invalid value for termination, acceptable values are %s, %s, or %s", routeapi.TLSTerminationEdge, routeapi.TLSTerminationPassthrough, routeapi.TLSTerminationReencrypt)
```
`NewNotSupportedError`

kargakis replied:
after the next rebase

deads2k replied:
> after the next rebase

`NewFieldValueNotSupported` in our level.

kargakis replied:
ok

deads2k commented:
Looking at the code, I think we actually wanted different types for these termination types.  We should remember to do that for v2.

deads2k commented:
@liggitt @smarterclayton this will leave us with potentially invalid route data in etcd.  The silver lining is that if that's the case, then their route wasn't really working in a predictable way before, so I don't feel terrible about breaking.

We could just make sure our controller didn't go insane and highlight the problem in `oc status`.

kargakis commented:
> Looking at the code, I think we actually wanted different types for these termination types. We should remember to do that for v2.

Added TODOs in the code regarding

liggitt commented:
> @liggitt @smarterclayton this will leave us with potentially invalid route data in etcd. The silver lining is that if that's the case, then their route wasn't really working in a predictable way before, so I don't feel terrible about breaking.
>
> We could just make sure our controller didn't go insane and highlight the problem in oc status.


Similar conversation going on in https://github.com/openshift/origin/pull/6037

kargakis commented:
> We could just make sure our controller didn't go insane and highlight the problem in oc status.

Opened https://github.com/openshift/origin/pull/6078

deads2k commented on pkg/route/api/validation/validation.go:
```diff
@@ -109,8 +109,8 @@ func validateTLS(route *routeapi.Route) fielderrors.ValidationErrorList {
-		msg := fmt.Sprintf("invalid value for termination, acceptable values are %s, %s, %s, or empty (no tls specified)", routeapi.TLSTerminationEdge, routeapi.TLSTerminationPassthrough, routeapi.TLSTerminationReencrypt)
-		result = append(result, fielderrors.NewFieldInvalid("termination", tls.Termination, msg))
+		msg := fmt.Sprintf("invalid value for termination, acceptable values are %s, %s, or %s", routeapi.TLSTerminationEdge, routeapi.TLSTerminationPassthrough, routeapi.TLSTerminationReencrypt)
+		result = append(result, fielderrors.NewFieldValueNotSupported("termination", tls.Termination, []string{msg}))
```
bad usage.  The string slice is for valid values.

kargakis replied:
rushed on this, thanks

On Wed, Nov 25, 2015 at 4:21 PM, David Eads <notifications@github.com>
wrote:

> In pkg/route/api/validation/validation.go
> <https://github.com/openshift/origin/pull/6076#discussion_r45878441>:
>
> > @@ -109,8 +109,8 @@ func validateTLS(route *routeapi.Route) fielderrors.ValidationErrorList {
> >  			result = append(result, fielderrors.NewFieldInvalid("destinationCACertificate", tls.DestinationCACertificate, "edge termination does not support destination certificates"))
> >  		}
> >  	default:
> > -		msg := fmt.Sprintf("invalid value for termination, acceptable values are %s, %s, %s, or empty (no tls specified)", routeapi.TLSTerminationEdge, routeapi.TLSTerminationPassthrough, routeapi.TLSTerminationReencrypt)
> > -		result = append(result, fielderrors.NewFieldInvalid("termination", tls.Termination, msg))
> > +		msg := fmt.Sprintf("invalid value for termination, acceptable values are %s, %s, or %s", routeapi.TLSTerminationEdge, routeapi.TLSTerminationPassthrough, routeapi.TLSTerminationReencrypt)
> > +		result = append(result, fielderrors.NewFieldValueNotSupported("termination", tls.Termination, []string{msg}))
>
> bad usage. The string slice is for valid values.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/openshift/origin/pull/6076/files#r45878441>.
>

jwforres commented:
@spadgett

kargakis commented:
@openshift/api-review any objections here?

smarterclayton commented:
You need to default the values if you're going to get stricter.  You can't break data in etcd without an incredibly good reason.

kargakis commented:
Added defaulting to edge termination

deads2k commented on pkg/route/api/v1/conversion.go:
```diff
@@ -13,6 +13,9 @@ func init() {
 		},
 		func(obj *TLSConfig) {
+			if len(obj.Termination) == 0 {
+				obj.Termination = TLSTerminationEdge
```
We aren't guaranteed to have the information required to make this work.  This changes behavior doesn't it?

kargakis replied:
Yes, it does. Actually I am not sure we should set any default for this. @smarterclayton how do you envision defaulting here?

cc: @pweil- @ramr @liggitt

deads2k replied:
> Yes, it does. Actually I am not sure we should set any default for this. @smarterclayton how do you envision defaulting here?

Basically, we have a problem where an existing object has a bunch of ignored data set, but there's no obvious indication that your intent (secure my route) failed.  The entire stanza has no reason to exist if its not used, so we probably should always have rejected it.

Dropping the entire stanza would maintain current behavior and allow us to tighten validation, but it would drop some (unused) data out of the object.  @smarterclayton would that satisfy?

pweil- replied:
anything stored currently with an empty termination type should be considered a non-secure route.  If we try to default to edge termination we're going to force the users to use a wildcard or default, invalid certificate unless we have cert information in the old route.  

Previously, an empty termination type on the route was consider non-secure anyway.  The router should ignore the certs with an empty termination field unless something has changed.

deads2k replied:
> Previously, an empty termination type on the route was consider non-secure anyway. The router should ignore the certs with an empty termination field unless something has changed.

We want to tighten validation to prevent users from creating borked routes with a bunch of useless values, but that would make existing data in etcd invalid, which brings its own problems with it.

pweil- replied:
sure, I'm down with that.  But since the router knows to DTRT I don't think it's that big of a deal.  If we really want to make it look like an unsecure route I'd say we unset the irrelevant data like you suggest.

kargakis replied:
What if they are set on purpose (plans for enabling https at a future point? Not sure how valid of a use case it is) and we silently remove them?

deads2k replied:
> What if they are set on purpose (plans for enabling https at a future point? Not sure how valid of a use case it is) and we silently remove them?

I think I'm ok with removing them.  It seems unlikely that I've chosen to use an openshift route as the sole location for storing my certs and keys, so it should be recoverable if that was their intent.  It would also be a pretty unusual intent: "I want this to be secure, but I'll start out my making it insecure"?

smarterclayton replied:
If we want to create a new type TerminationNone that would be more accurate
(and default to that).  However, a user who creates and adds certs almost
certainly wants termination, not "none"

On Mon, Dec 14, 2015 at 9:07 AM, David Eads <notifications@github.com>
wrote:

> In pkg/route/api/v1/conversion.go
> <https://github.com/openshift/origin/pull/6076#discussion_r47500560>:
>
> > @@ -13,6 +13,9 @@ func init() {
> >  			obj.To.Kind = "Service"
> >  		},
> >  		func(obj *TLSConfig) {
> > +			if len(obj.Termination) == 0 {
> > +				obj.Termination = TLSTerminationEdge
>
> What if they are set on purpose (plans for enabling https at a future
> point? Not sure how valid of a use case it is) and we silently remove them?
>
> I think I'm ok with removing them. It seems unlikely that I've chosen to
> use an openshift route as the sole location for storing my certs and keys,
> so it should be recoverable if that was their intent. It would also be a
> pretty unusual intent: "I want this to be secure, but I'll start out my
> making it insecure"?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/openshift/origin/pull/6076/files#r47500560>.
>

deads2k replied:
> If we want to create a new type TerminationNone that would be more accurate
> (and default to that).  However, a user who creates and adds certs almost
> certainly wants termination, not "none"

I'm not sure it makes sense to have a `TLSConfig` object, whose sole purpose is provide TLS termination information with a flag on it that means, "don't use the TLSConfig I provided".

smarterclayton replied:
In which case defaulting to Terminate is the right thing.

On Mon, Dec 14, 2015 at 10:12 AM, David Eads <notifications@github.com>
wrote:

> In pkg/route/api/v1/conversion.go
> <https://github.com/openshift/origin/pull/6076#discussion_r47508030>:
>
> > @@ -13,6 +13,9 @@ func init() {
> >  			obj.To.Kind = "Service"
> >  		},
> >  		func(obj *TLSConfig) {
> > +			if len(obj.Termination) == 0 {
> > +				obj.Termination = TLSTerminationEdge
>
> If we want to create a new type TerminationNone that would be more accurate
> (and default to that). However, a user who creates and adds certs almost
> certainly wants termination, not "none"
>
> I'm not sure it makes sense to have a TLSConfig object, whose sole
> purpose is provide TLS termination information with a flag on it that
> means, "don't use the TLSConfig I provided".
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/openshift/origin/pull/6076/files#r47508030>.
>

liggitt replied:
removing the TLS data in conversion means we can't catch it in validation. making an object in this state fail validation would mean it couldn't be updated in etcd without setting the termination type or removing the cert/key data... what would that break?

deads2k replied:
> removing the TLS data in conversion means we can't catch it in validation. making an object in this state fail validation would mean it couldn't be updated in etcd without setting the termination type or removing the cert/key data... what would that break?

Right, had a conversation in irc.  If we tighten validation, do no defaulting, and create a special path for `validateUpdate` for `routes/status`, nothing breaks, but @smarterclayton wasn't convinced.

smarterclayton replied:
I don't want to have different special paths.  This problem doesn't bother
me enough to justify having a split brain on validation.

On Mon, Dec 14, 2015 at 10:22 AM, David Eads <notifications@github.com>
wrote:

> In pkg/route/api/v1/conversion.go
> <https://github.com/openshift/origin/pull/6076#discussion_r47509346>:
>
> > @@ -13,6 +13,9 @@ func init() {
> >  			obj.To.Kind = "Service"
> >  		},
> >  		func(obj *TLSConfig) {
> > +			if len(obj.Termination) == 0 {
> > +				obj.Termination = TLSTerminationEdge
>
> removing the TLS data in conversion means we can't catch it in validation.
> making an object in this state fail validation would mean it couldn't be
> updated in etcd without setting the termination type or removing the
> cert/key data... what would that break?
>
> Right, had a conversation in irc. If we tighten validation, do no
> defaulting, and create a special path for validateUpdate for routes/status,
> nothing breaks, but @smarterclayton <https://github.com/smarterclayton>
> wasn't convinced.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/openshift/origin/pull/6076/files#r47509346>.
>

liggitt replied:
ValidateRouteStatusUpdate doesn't validate route spec

deads2k replied:
> ValidateRouteStatusUpdate doesn't validate route spec


already split brained? :)

liggitt replied:
yes, and if we tighten validation, we should add a comment to `ValidateRouteStatusUpdate` noting that we should not call ValidateRouteUpdate or we'll break status updates on old invalid routes

kargakis replied:
I don't see what validation on `routes/status` got to do with this. 

If http will continue to function after defaulting an empty termination field to edge, why not just do it?

deads2k replied:
> I don't see what validation on routes/status got to do with this.
> 
> If http will continue to function after defaulting an empty termination field to edge, why not just do it?

A route that contains a `TLSConfig` *must* have an `Termination` value or it doesn't make sense.  We want validation to enforce that so that users will be smacked when they try to create such an object.

That change makes some objects stored in etcd invalid.

Making objects in etcd invalid presents problems for controllers that try to update particular fields.  If the controller doesn't fix the "broken" field, then validation can prevent updates by the controllers which can make controllers angry.

Lucky for us, `routes/status` uses different validation, which does not require that the `Route` itself be valid.  This allows us to update the validation without changing defaulting rules and not break our controllers.

kargakis replied:
> when they try to create such an object

Is [route status validation](https://github.com/openshift/origin/blob/3cf7f3cccd5b7b95448568ba8e2b3c81cc6ab9af/pkg/route/registry/route/strategy.go#L99-L101) running when creating a route or only when a route is updated?

smarterclayton replied:
It doesn't matter what we want here - if we shipped an API that
accepted an empty value and behaved one way, then realized it need to
be set for clarity, a default value is exactly how we solve that.

deads2k replied:
> It doesn't matter what we want here - if we shipped an API that
> accepted an empty value and behaved one way, then realized it need to
> be set for clarity, a default value is exactly how we solve that.

Spoke with @smarterclayton in person.  Here's the net:
 1. We can add validation to reject `TLS.Termination==""`
 1. If the `TLSConfig` does not have a `Termination` and does not have a `DestinationCACertificate` and `validateInsecureEdgeTerminationPolicy` would return ok, then default `TLS.Termination` to edge.  This is basically, "if the `TLSConfig` has a shot at working for edge termination.

@smarterclayton one thing I forgot to ask.  If the values are present, but broken, will the unencrypted route still work?

smarterclayton replied:
Yes, that should work.

kargakis commented:
updated as per the latest comments

deads2k commented on pkg/route/api/v1/conversion.go:
```diff
@@ -13,6 +13,10 @@ func init() {
 			obj.To.Kind = "Service"
 		},
 		func(obj *TLSConfig) {
+			err := ValidateInsecureEdgeTerminationPolicy(obj)
```
This is a different package.

kargakis replied:
doh. Now that I actually fixed this import, it requires conversion from v1 to internal in order for the validation function to work. Should I proceed with converting just to validate, duplicate the validation function to work against the versioned object, or isn't this validation step really necessary?

kargakis replied:
Existing routes that won't pass `ValidateInsecureEdgeTerminationPolicy` are already broken, right? It seems to me that this step isn't really necessary here.

deads2k commented on pkg/route/api/validation/validation.go:
```diff
@@ -67,14 +67,14 @@ func ValidateRouteStatusUpdate(route *routeapi.Route, older *routeapi.Route) fie
 	return allErrs
```
Add a comment to this method indicating that it can't call `ValidateRouteUpdate` or we'll risk breaking

kargakis replied:
ok

deads2k commented:
Minor comments, lgtm otherwise.

deads2k commented:
lgtm try merging once the merge queue works better

[test]

deads2k commented:
[merge]

liggitt commented:
test failures:
```
--- FAIL: TestValidateTLS (0.00s)
	validation_test.go:367: Test case No TLS Termination expected 0 error(s), got 1. [termination: unsupported value '', Details: supported values: edge, passthrough, reencrypt]
--- FAIL: TestValidateNoTLSInsecureEdgeTerminationPolicy (0.00s)
	validation_test.go:505: Test case for no termination with insecure=None got 2 errors where none were expected. [termination: unsupported value '', Details: supported values: edge, passthrough, reencrypt InsecureEdgeTerminationPolicy: invalid value 'None', Details: InsecureEdgeTerminationPolicy is only allowed for edge-terminated routes]
	validation_test.go:505: Test case for no termination with insecure=Allow got 2 errors where none were expected. [termination: unsupported value '', Details: supported values: edge, passthrough, reencrypt InsecureEdgeTerminationPolicy: invalid value 'Allow', Details: InsecureEdgeTerminationPolicy is only allowed for edge-terminated routes]
	validation_test.go:505: Test case for no termination with insecure=Redirect got 2 errors where none were expected. [termination: unsupported value '', Details: supported values: edge, passthrough, reencrypt InsecureEdgeTerminationPolicy: invalid value 'Redirect', Details: InsecureEdgeTerminationPolicy is only allowed for edge-terminated routes]
	validation_test.go:505: Test case for no termination with insecure=support HTTPsec got 2 errors where none were expected. [termination: unsupported value '', Details: supported values: edge, passthrough, reencrypt InsecureEdgeTerminationPolicy: invalid value 'support HTTPsec', Details: InsecureEdgeTerminationPolicy is only allowed for edge-terminated routes]
	validation_test.go:505: Test case for no termination with insecure=or maybe HSTS got 2 errors where none were expected. [termination: unsupported value '', Details: supported values: edge, passthrough, reencrypt InsecureEdgeTerminationPolicy: invalid value 'or maybe HSTS', Details: InsecureEdgeTerminationPolicy is only allowed for edge-terminated routes]
FAIL
coverage: 87.5% of statements
FAIL	github.com/openshift/origin/pkg/route/api/validation	0.050s
ok  	github.com/openshift/origin/pkg/route/controller/allocation	0.052s	coverage: 83.3% of statements
ok  	github.com/openshift/origin/pkg/route/generator	0.041s	coverage: 73.9% of statements
--- FAIL: TestMissingTLSTerminationType (0.02s)
	analysis_test.go:51: expected 1 markers, got 0
FAIL
coverage: 94.4% of statements
FAIL	github.com/openshift/origin/pkg/route/graph/analysis	0.083s
```

kargakis commented:
Updated validation tests and removed an oc status test since the fixture is now broken but left the oc status error behind since the suggestion is useful.

kargakis commented on pkg/route/api/v1/conversion.go:
```diff
@@ -13,6 +13,9 @@ func init() {
 		},
 		func(obj *TLSConfig) {
+			if len(obj.Termination) == 0 && len(obj.DestinationCACertificate) == 0 {
+				obj.Termination = TLSTerminationEdge
```
@deads2k how about initializing to Re-encrypt when there is a DestinationCACertificate?

deads2k replied:
> @deads2k how about initializing to Re-encrypt when there is a DestinationCACertificate?

I'm less certain about that.  @pweil-  may know.  Seems like re-encryption would require addition keys.

pweil- replied:
Yes, re-encrypt would require the destination ca certificate.  You could potentially not have the cert/ca cert on the route if you're terminating with a wildcard cert at the edge.
